### PR TITLE
CLI: Resolves #6478: Added note count indicator per notebook

### LINKED ISSUE #6478: Add note counts indicators per notebook to the CLI/Terminal app
The desktop app has useful information displayed on the notebook list to show how many notes are contained within it.
e.g. 
image

The goal of this would be to extend this functionality to the terminal application.

- Counts of notes within notebooks should be displayed inline in the notebook pane as an indicator after the notebook entry
- Consider adding a config option to enable or disable the functionality - unlike the desktop application which can be modified with CSS, the terminal app doesn't have such user accessible UI customisation but the addition of this might cause undue clutter on lower resolution displays where the terminal app might be used

### PR DESCRIPTION
**Pull request by eduebernal**
Resolves #6478 

Created a config option to enable or disable note count in the notebook list gui.

image

https://user-images.githubusercontent.com/92958867/169619963-5bff6f82-9c22-41fd-8955-e9d507a91430.mov

I'm not sure if these changes are testable, but if they are I would appreciate the help of any of you on how to approach those tests.

### CODE REVIEW

**Comment by laurent22** [comment]
The code seems ok overall although I'll have a few comments. @Daeraxa is that roughly what you had in mind in terms of features?

**Comment by Daeraxa** [comment]
That is pretty much exactly what I'm thinking in terms of features yes, maybe if/when we come to colour customisation (https://github.com/laurent22/joplin/issues/6481) then we could add an element to knock it down a shade.

**Comment by eduebernal** [comment]
Hi! Just following up this PR, hope you aren't too busy. I'm available to work on any comments or fixes you propose @laurent22.

File: packages/app-cli/app/gui/FolderListWidget.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const ListWidget = require('tkwidgets/ListWidget.js');
 const _ = require('@joplin/lib/locale')._;
 
 class FolderListWidget extends ListWidget {
-	constructor() {
+	constructor(setting) {
```
Comment (laurent22) [inline packages/app-cli/app/gui/FolderListWidget.js]: I don't think we need to pass the Setting class here, just include it directly in this file.

File: packages/lib/models/Setting.ts [hunk 2]
```diff
@@ -991,6 +991,14 @@ class Setting extends BaseModel {
 				storage: SettingStorage.File,
 			},
 			'folders.sortOrder.reverse': { value: false, type: SettingItemType.Bool, storage: SettingStorage.File, isGlobal: true, public: true, label: () => _('Reverse sort order'), appTypes: [AppType.Cli] },
+			'folders.showNoteCount': {
```
Comment (laurent22) [inline packages/lib/models/Setting.ts]: We already have the setting `showNoteCounts` so please use that. Just enable it for cli too, under appTypes

File: readme/terminal.md [hunk 3]
```diff
@@ -488,6 +488,10 @@ The following commands are available in [command-line mode](#command-line-mode):
 	                                   Type: bool.
 	                                   Default: false
 
+	    folders.showNoteCount          Show note count in folder list.
+	                                   Type: bool.
+	                                   Default: true
+
 	    trackLocation                  Save geo-location with notes.
```
Comment (laurent22) [inline readme/terminal.md]: Please don't update this file, that will be done automatically later on

**Comment by tessus** [comment]
The question is how you count the notes. The way you do it is different than what the desktop version does.

e.g. If you have a folder and subfolder:

There is one note in `Folder` and 2 notes in `Subfolder`:

# Terminal expanded (I don't think there's a way to collapse in the cli app)

```
Folder            3
   Subfolder      2
```

# Desktop expanded

```
Folder            1
   Subfolder      2
```

# Desktop collapsed

```
Folder            3
```

**Comment by eduebernal** [comment]
@tessus  You are right, I changed the logic to match the desktop app. As far as I can tell, you can't create subfolders with the cli app by itself. I created some folders and subfolders with the desktop app and synced them to the file system in order to see that behavior.

image

File: packages/lib/models/Setting.ts (hunk 2 above)
Comment (eduebernal) [reply packages/lib/models/Setting.ts]: Done [fdf2e3b](https://github.com/laurent22/joplin/pull/6526/commits/fdf2e3bf11119e2aca948866b715c2edbe3522b5)

File: packages/app-cli/app/gui/FolderListWidget.js (hunk 1 above)
Comment (eduebernal) [reply packages/app-cli/app/gui/FolderListWidget.js]: Done [3785a1c](https://github.com/laurent22/joplin/pull/6526/commits/3785a1cd38ba39b1dc6290af6341eab6470dc9d6)

File: readme/terminal.md (hunk 3 above)
Comment (eduebernal) [reply readme/terminal.md]: Ok! [60d2b24](https://github.com/laurent22/joplin/pull/6526/commits/60d2b24415d3b67727d5e5a26ad86ee21eeb3e41)

File: packages/app-cli/app/gui/FolderListWidget.js [hunk 4]
```diff
@@ -25,6 +26,18 @@ class FolderListWidget extends ListWidget {
 				output.push('-'.repeat(this.innerWidth));
 			} else if (item.type_ === Folder.modelType()) {
 				output.push(' '.repeat(this.folderDepth(this.folders, item.id)) + Folder.displayTitle(item));
+				if (Setting.value('showNoteCounts')) {
```
Comment (laurent22) [inline packages/app-cli/app/gui/FolderListWidget.js:29]: By the way, is the folder list correctly update when the user change this setting value? using `:config showNoteCounts 0`

Comment (eduebernal) [reply packages/app-cli/app/gui/FolderListWidget.js:29]: Yeah, using 0 updates that setting to _false_.

Comment (laurent22) [reply packages/app-cli/app/gui/FolderListWidget.js:29]: I mean does it correctly hide/show the note count when you set showNoteCounts to 0/1? (without having to restart the app)

Comment (eduebernal) [reply packages/app-cli/app/gui/FolderListWidget.js:29]: Yes it does.

https://user-images.githubusercontent.com/92958867/174653691-6198fa1a-9b4a-4b4d-97dc-424bec0af583.mov

**Pull request merged**